File: Components/LiveTranslation/LiveTranslation.cs
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.ComponentModel.Composition;
using System.Linq;
using System.Text;
using System.Text.RegularExpressions;

using BeerViewer.Network;
using BeerViewer.Modules;

namespace LiveTranslation
{
	[Export(typeof(IBeerComponent))]
	[ExportMetadata("Guid", "91A946EE-1FDC-4C5D-8E1A-F3ECE38B42AA")]
	[ExportMetadata("Name", "LiveTranslation")]
	[ExportMetadata("Description", "In-game translation for korean user")]
	[ExportMetadata("Version", "1.1")]
	[ExportMetadata("Author", "WolfgangKurz")]
	[ExportMetadata("Permission", BeerComponentPermission.CP_NETWORK_MODIFIABLE)]
	public class LiveTranslation : IBeerComponent
	{
		public string Name => "LiveTranslation";

		public void Initialize()
		{
			Logger.Log(
				"LiveTranslation ver " +
					this.GetType()
						.GetCustomAttributes(false)
						.Where(x => x.GetType() == typeof(ExportMetadataAttribute))
						.Select(x => x as ExportMetadataAttribute)
						.FirstOrDefault(x => x.Name == "Version")
						.Value
			);

			var proxy = Proxy.Instance;
			proxy.RegisterModifiable((s, b) =>
			{
				if (s.Request.PathAndQuery == "/kcsapi/api_start2/getData")
				{
					var utf8 = Encoding.UTF8;
					var src = utf8.GetString(b);
					/*
					src = new Regex("(\"api_map_bgm\"):\\[[0-9]+,[0-9]+\\]")
						.Replace(src, "$1:[67,67]");

					src = new Regex("(\"api_boss_bgm\"):\\[[0-9]+,[0-9]+\\]")
						.Replace(src, "$1:[17,17]");
					*/
					var list = i18n.Current.Table
						.OrderByDescending(x => x.Key.Length);
					foreach (var pair in list)
					{
						var key = pair.Key;
						src = src.Replace(
							$"\"{EncodeUnicode(key)}\"",
							$"\"{EncodeUnicode(pair.Value)}\""
						);
					}

					b = utf8.GetBytes(src);
				}
				/*
				else if (s.Request.PathAndQuery == "/kcsapi/api_port/port")
				{
					var utf8 = Encoding.UTF8;
					var src = utf8.GetString(b);

					src = new Regex("(\"api_p_bgm_id\"):[0-9]+,")
						.Replace(src, "$1:125,");

					b = utf8.GetBytes(src);
				}
				*/

				return b;
			});

			;
		}

		public void ReadyFailed()
		{
			// Nothing to do
		}

		private string EncodeUnicode(string value)
		{
			var sb = new StringBuilder();
			foreach (var c in value)
			{
				if (c > 127)
					sb.Append("\\u" + ((int)c).ToString("x4"));
				else
					sb.Append(c);
			}
			return sb.ToString();
		}
	}
}

File: WindowFrame/source/Backend/LiveTranslation.ts
import Proxy from "../Proxy/Proxy";
import i18n from "../i18n/i18n";
import Unicode from "../Base/Unicode";

class LiveTranslation {
	private _Initialized: boolean = false;
	private list: { key: string, value: string }[];

	constructor() {
		let ko = i18n.get("ko");
		if (!ko) ko = "";

		this.list = ko.replace(/\r/g, "").split("\n")
			.filter(x => x.length > 0)
			.map(x => x.split("\t").filter(y => y.length > 0))
			.filter(x => x.length == 2)
			.map(x => ({
				key: x[0],
				value: x[1]
			}))
			.sort((a, b) => a.key.length < b.key.length ? -1 : a.key.length > b.key.length ? 1 : 0);
	}

	public Init(): void {
		if (this._Initialized) return;
		this._Initialized = true;

		Proxy.Instance.RegisterModifiable("/kcsapi/api_start2/getData", (req, resp) => {
			let body = resp.toString();
			this.list.forEach(x => {
				const a = Unicode.Escape(x.key);
				const b = Unicode.Escape(x.value);
				if (a === b) return;

				while (body.includes(`"${a}"`))
					body = body.replace(`"${a}"`, `"${b}"`);
			});
			return Buffer.from(body);
		});
	}
}
export default new LiveTranslation();